[processing] Fix running commands on layers with open options or credential options

nyalldawson opened the pull request:
(temporarily includes https://github.com/qgis/QGIS/pull/57868)

These improvements fix generation of GDAL commands when the input layers have open options specified (translating these to -oo arguments wherever the GDAL command supports these), or when credential options are specified for the QGIS layer (translating to --config arguments).

In addition to other fixes, this permits the GDAL tools to be run directly on layers loaded from eg a S3 bucket via GDAL vsis3.

nyalldawson committed:
Cleanup API for GDAL processing provider uri conversion

nyalldawson committed:
minor cleanup

nyalldawson committed:
Make getOgrCompatibleSource API more flexible

nyalldawson committed:
Pass vector layer open options over to GDAL command

Ensures that processing tools which use vector layers with open
options work correctly, respecting the layer's open option settings

nyalldawson committed:
Correctly handle open options for raster GDAL algorithm inputs

And handoff to GDAL command when command supports them

nyalldawson committed:
Fix running GDAL algorithms on cloud storage layers

Correctly generate commands when run on eg /vsis3/ layer sources,
and translate credential options to "--config KEY VALUE" arguments
for GDAL commands

nyalldawson committed:
Regenerate some GDAL output reference files

We now pass on the FORCE_SRS_DETECTION=YES open option which we
were previously dropping, so the results for these commands is
slightly different

Also drop test for older GEOS versions, it's covered by a test
for more recent versions and we have extensive test covering
the GDAL command line for this tool anyway.

alexbruy commented on python/plugins/processing/algs/gdal/GridAverage.py:
```diff
@@ -180,6 +182,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Support for open options in this tool was added in GDAL 3.7. Maybe we should check GDAL version here to support old platforms?

alexbruy commented on python/plugins/processing/algs/gdal/GridDataMetrics.py:
```diff
@@ -194,6 +196,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Same as above

alexbruy commented on python/plugins/processing/algs/gdal/GridInverseDistance.py:
```diff
@@ -203,6 +205,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Same as above

alexbruy commented on python/plugins/processing/algs/gdal/GridInverseDistanceNearestNeighbor.py:
```diff
@@ -188,6 +190,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Also here

alexbruy commented on python/plugins/processing/algs/gdal/GridLinear.py:
```diff
@@ -160,6 +162,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
And here

alexbruy commented on python/plugins/processing/algs/gdal/GridNearestNeighbor.py:
```diff
@@ -174,6 +174,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Same as above

alexbruy commented on python/plugins/processing/algs/gdal/rasterize.py:
```diff
@@ -237,6 +235,12 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('-of')
         arguments.append(output_format)
 
+        if input_details.open_options:
```
Support for open options was added in GDAL 3.7, maybe add a check for GDAL version here?

alexbruy commented on python/plugins/processing/algs/gdal/rasterize_over.py:
```diff
@@ -119,8 +119,14 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
 
-        arguments.append(ogrLayer)
-        arguments.append(inLayer.source())
+        if input_details.open_options:
```
Support for open options was added in GDAL 3.7, maybe check GDAL version here?

alexbruy commented on python/plugins/processing/algs/gdal/rasterize_over_fixed_value.py:
```diff
@@ -96,29 +93,37 @@ def commandName(self):
         if self.parameterAsBool(parameters, self.ADD, context):
             arguments.append('-add')
 
+        if input_details.open_options:
```
Same as above

nyalldawson replied:
Thanks, done

alexbruy reviewed (approved):
Looks good to me. Just not sure if we need a check for GDAL version if a few algorithms where support for open options was added in GDAL 3.7.

nyalldawson committed:
Raise exceptions if open options not supported by GDAL version

nyalldawson committed:
Only run gdal_grid, gdal_rasterize tests on GDAL 3.7+

The inputs we use for these run tests have open options, and now
that we are respecting these and passing them to the GDAL command
we can't run these tests on the earlier GDAL versions which
didn't support open options.

The commands generated by the corresponding algorithms are already
well tested and still run on the earlier GDAL releases, it's just
the actual run test which will be skipped.

agiudiceandrea commented:
Hi @nyalldawson, please have a look at the regression reported at https://github.com/qgis/QGIS/issues/61144.
